third_party/amd/include/TritonAMDGPUTransforms/SchedulerMachineModel.h

```diff
@@ -0,0 +1,489 @@
+  best map to.
+  For example, mfma_16x16x16 on MI300X
+  - Takes 4 cycles to issue.
+  - Also keeps the mfma pipe busy for an additional 12 cycles
+  Therefore 4 back-to-back mfmas will take 4*16=64 cycles since
+  they're each taking up 16 cycles of the mfma pipe.
+  And 1 mfma, 1 lds op, 1 mfma, 1 other op only takes 4*16=32 cycles
+  since the mfmas took 16 cycles but the other ops mapped to different
+  hardware pipes and could be co-executed for free.
+
+  This complexity of modeling co-execution of ops is necessary
+  for determining, e.g., at the top of a loop, how many LoadOps
+  can be grouped with how many DotOps, since the LoadOps need
+  to increment their addresses and then can co-execute with
+  the DotOps.
+*/
+struct MachineModelOpProperties {
+
+  MachineModelOpProperties(MachineModelResourcePipe pipe, int32_t seqBusy,
+                           StringRef n, int32_t pipeBusyAfterSeq = 0)
+      : resourcePipe(pipe), cyclesSeqBusy(seqBusy),
+        cyclesPipeBusyAfterSeq(pipeBusyAfterSeq), name(n) {}
+
+  // To which resource pipe does this op map.
+  MachineModelResourcePipe resourcePipe;
+
+  // Op blocks all other ops from issuing.
+  int32_t cyclesSeqBusy;
+
+  /*
+    How long to wait before issuing another op to same
+    resource pipe even after sequencer issues.
+    E.g. mfma "takes" 16 cycles, but this is broken into 4 cycles that the
+    sequencer is busy issuing the mfma, and 12 more cycles that the
+    mfma pipe is still busy but other pipes can be used.
+  */
+  int32_t cyclesPipeBusyAfterSeq;
+
+  // To be used for debugging, e.g. saying that op was
+  // assumed to map to 16 v_pk_add_fp32 instructions.
+  StringRef name;
+};
+
+/*
+  Abstract base class for querying op properties based on GPU generation.
+  Each new generation of hardware inherits from the previous,
+  therefore each new geneation only needs to override ops with new properties.
+  Most ops we don't yet care about, so most ops will fall back to whatever
+  the most common instruction is, e.g., 4 cycles of valu.
+*/
+struct MachineModel {
+
+  virtual MachineModelOpProperties getOpProperties(Operation *op) = 0;
+
+  /*
+    Models how many cycles does it take between completing issuing
+    a memory op (seqBusy done) and when the data is ready.
+    This is currently modeled based on pipe and not on op since
+    it is assumed, e.g., that ds_read_b32 has same latency
+    as ds_read_b64.
+    It is also assumed that read and writes to a pipe are the same.
+  */
+  virtual int32_t getDataLatency(MachineModelResourcePipe pipe) {
+    switch (pipe) {
+    case Lds:
+      return 80;
+    case Global:
+      // 1M cycles is dummy value to push buffer loads as early as possible.
+      return 1000000;
+    default:
+      return 0;
+    }
+  };
+
+  virtual ~MachineModel() = default;
+};
+
+// MI250
+struct MachineModelGFX90A : MachineModel {
+  MachineModelOpProperties getOpProperties(Operation *op) {
```

Comment:
`MachineModelOpProperties getOpProperties(Operation *op) override {`

The same below